test/sslapitest.c

```diff
@@ -9229,6 +9229,109 @@ static int test_pluggable_group(int idx)
+    X509_NAME *name = NULL;
+    BIO *keybio = NULL, *certbio = NULL;
+    int ret = 1;
+
+    if (!TEST_ptr(evpctx)
+        || !TEST_true(EVP_PKEY_keygen_init(evpctx))
+        || !TEST_true(EVP_PKEY_generate(evpctx, &pkey))
+        || !TEST_ptr(pkey)
+        || !TEST_ptr(x509)
+        || !TEST_true(ASN1_INTEGER_set(X509_get_serialNumber(x509), 1))
+        || !TEST_true(X509_gmtime_adj(X509_getm_notBefore(x509), 0))
+        || !TEST_true(X509_gmtime_adj(X509_getm_notAfter(x509), 31536000L))
+        || !TEST_true(X509_set_pubkey(x509, pkey))
+        || !TEST_ptr(name = X509_get_subject_name(x509))
+        || !TEST_true(X509_NAME_add_entry_by_txt(name, "C",  MBSTRING_ASC,
+                           (unsigned char *)"CH", -1, -1, 0))
+        || !TEST_true(X509_NAME_add_entry_by_txt(name, "O",  MBSTRING_ASC,
+                           (unsigned char *)"test.org", -1, -1, 0))
+        || !TEST_true(X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC,
+                           (unsigned char *)"localhost", -1, -1, 0))
+        || !TEST_true(X509_set_issuer_name(x509, name))
+        || !TEST_true(X509_sign(x509, pkey, EVP_sha1()))
+        || !TEST_ptr(keybio = BIO_new_file(privkeyfilename, "wb"))
+        || !TEST_true(PEM_write_bio_PrivateKey(keybio, pkey, NULL, NULL, 0, NULL, NULL))
+        || !TEST_ptr(certbio = BIO_new_file(certfilename, "wb"))
+        || !TEST_true(PEM_write_bio_X509(certbio, x509)))
+        ret = 0;
+
+    EVP_PKEY_free(pkey);
+    X509_free(x509);
+    EVP_PKEY_CTX_free(evpctx);
+    BIO_free(keybio);
+    BIO_free(certbio);
+    return ret;
+}
+
+/*
+ * Test that signature algorithms loaded via the provider interface can
+ * correctly establish a TLS (1.3) connection.
+ * Test 0: Signature algorithm with built-in hashing functionality: "xorhmacsig"
+ * Test 1: Signature algorithm using external SHA2 hashing: "xorhmacsha2sig"
+ */
+static int test_pluggable_signature(int idx)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    OSSL_PROVIDER *tlsprov = OSSL_PROVIDER_load(libctx, "tls-provider");
+    OSSL_PROVIDER *defaultprov = OSSL_PROVIDER_load(libctx, "default");
+    char *certfilename = "tls-prov-cert.pem";
+    char *privkeyfilename = "tls-prov-key.pem";
+
+    /* create key and certificate for the different algorithm types */
+    if (!TEST_ptr(tlsprov)
+        || !TEST_true(create_cert_key(idx, certfilename, privkeyfilename)))
+        goto end;
+
+    if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_3_VERSION,
+                                       TLS1_3_VERSION,
+                                       &sctx, &cctx, certfilename, privkeyfilename))
+            || !TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                             NULL, NULL)))
+        goto end;
```

Comment:
So this is the part that caused the addition of so much boilerplate in `tls-provider.c`...

All that churn could easily be avoided by avoiding files, which is perfect possible.  Try this:

```suggestion
    EVP_PKEY_CTX * evpctx = NULL;
    EVP_PKEY *pkey = NULL;
    X509 *x509 = NULL;

    /* create key and certificate for the different algorithm types */
    if (!TEST_ptr(tlsprov)
        || !TEST_ptr(evpctx = EVP_PKEY_CTX_new_from_name(libctx,
                                                         (idx == 0)
                                                         ? "xorhmacsig"
                                                         : "xorhmacsha2sig",
                                                         NULL))
        || !TEST_int_gt(EVP_PKEY_keygen_init(evpctx), 0)
        || !TEST_int_gt(EVP_PKEY_generate(evpctx, &pkey), 0)
        || !TEST_ptr(x509 = create_cert(pkey)))
        goto end;

    if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
                                       TLS_client_method(),
                                       TLS1_3_VERSION,
                                       TLS1_3_VERSION,
                                       &sctx, &cctx, NULL, NULL))
        || !TEST_int_gt(SSL_CTX_use_certificate(sctx, x509), 0)
        || !TEST_int_gt(SSL_CTX_use_PrivateKey(sctx, pkey), 0)
        || !TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
                                         NULL, NULL)))
        goto end;
```

You'll have to rename and edit `create_ctert_key()` a bit, so it ends up having this signature, and doesn't write any files:

``` C
static X509 *create_cert(EVP_PKEY *pkey);
```

I believe that given this change, you can remove *all* the encoder / decoder code from `tls-provider.c`

Reply:
Thanks very much for this suggestion, @levitte ! If only I knew that a few months earlier :-/. But before following through on this proposal, this question: Is there _any_ benefit from having all this code, e.g., to retain/use it to (allow others to :) write more test cases for other "provider corner cases"?

Reply:
Thinking of @mattcaswell's comments on the amount of boilerplate, I think we'll want to see if we can create some standard stuff to make such things easier, so I think it's safe to throw away.

And mind you, I know the feeling of cutting away code that you've worked on hard.  Twist of knife kinda feeling.
The takeaway is, however, that it served as food for thought.

Reply:
Thanks. https://www.goodreads.com/quotes/410913-everyone-in-life-has-a-purpose-even-if-it-s-to. Will remove.

Reply:
@levitte: I'm afraid  there's a glitch: While it's true that one can remove all file write/read logic, it seems to be not right that this means encode/decode logic can be left away: If I'm not mistaken (?) one still needs to associate the EVP_PKEY-under-test with a corresponding X509 structure, right (or is there another way to get a cert for the TLS connection)? 

Using `X509_PUBKEY_set` for this, the following code gets triggered: https://github.com/openssl/openssl/blob/accd3bdd11bd4a69fdba42bbeead28945fe50e56/crypto/x509/x_pubkey.c#L345-L361 which (IMO logically) needs at least an SPKI encoder.

All that's needed to trigger this (mis)behaviour is disable the encode/decode code tables in the `tls_prov_query` function and the tests fail:

```       # ERROR: (bool) 'X509_set_pubkey(x509, pkey) == true' failed @ test/sslapitest.c:9249
        # false
        # ERROR: (ptr) 'x509 = create_cert(pkey) != NULL' failed @ test/sslapitest.c:9295
        # 0x0
        # 809B9019B27F0000:error:1D800065:ENCODER routines:OSSL_ENCODER_to_bio:reason(101):crypto/encode_decode/encoder_lib.c:55:No encoders were found. For standard encoders you need at least one of the default or base providers available. Did you forget to load them?
        # 809B9019B27F0000:error:0580006F:x509 certificate routines:X509_PUBKEY_set:unsupported algorithm:crypto/x509/x_pubkey.c:364:
```

Reply:
Oh, right.... I had forgotten about those internal conversions that use decoders and encoders

Reply:
So I guess we can close this conversation then?

Reply:
Yes